src/jupyterlabcontroller/services/lab.py

```diff
@@ -69,175 +206,339 @@ async def create_lab(
+        """Iterator over the events for a user.
 
-        # Build the objects that make up the user's lab.
-        objects = self._builder.build_lab(
-            user=user,
-            lab=lab,
-            image=image,
-            secrets=secret_data,
-            pull_secret=pull_secret,
-        )
+        Parameters
+        ----------
+        username
+            Username for which to retrieve events.
 
-        # Create the lab objects in Kubernetes.
-        self._logger.info("Creating new lab")
-        await self._storage.create(objects)
-        await self._lab_state.publish_pod_creation(
-            username, "Created Kubernetes objects for user lab", 30
-        )
+        Yields
+        ------
+        Event
+            Events for that user until a completion or failure event is seen.
 
-        # Return the URL where the lab will be listening after it starts.
-        return self._builder.build_internal_url(user.username, lab.env)
+        Raises
+        ------
+        UnknownUserError
+            Raised if there is no event stream for this user.
+        """
+        if username not in self._labs:
+            raise UnknownUserError(f"Unknown user {username}")
 
-    async def delete_lab(self, username: str) -> None:
-        """Delete the lab environment for the given user.
+        async def iterator() -> AsyncIterator[ServerSentEvent]:
+            async for event in self._labs[username].events:
+                yield event.to_sse()
+
+        return iterator()
+
+    async def get_lab_state(self, username: str) -> UserLabState | None:
+        """Get lab state for a user.
+
+        This method underlies the API called by JupyterHub to track whether
+        the user's lab still exists. We want to update that state after
+        Kubernetes changes faster than our reconciliation cycle, so ask
+        Kubernetes directly for the pod phase each time we're asked for the
+        lab state.
 
         Parameters
         ----------
         username
-            Username whose environment should be deleted.
+            Username to retrieve lab state for.
+
+        Returns
+        -------
+        UserLabState or None
+            Lab state for that user, or `None` if that user doesn't have a
+            lab.
 
         Raises
         ------
-        KubernetesError
-            Raised if a Kubernetes error prevented lab deletion.
         UnknownUserError
-            Raised if no lab currently exists for this user.
+            Raised if the given user has no lab.
         """
-        callback = partial(self._delete_lab_and_namespace, username, 25, 100)
-        await self._lab_state.stop_lab(username, callback)
+        if username not in self._labs:
+            return None
 
-    async def _delete_lab_and_namespace(
-        self, username: str, start_progress: int, end_progress: int
+        # Grab a copy of the lab state, since we're going to make async calls
```

Comment:
Yeah, this is what I meant about the lab_monitor above.

Reply:
Right, here we have to grab a copy because we're using await. The other code is set up so that it should never need to call await in the critical section.